### PR TITLE
Conditional facts and theories - support SQL2008 in testing

### LINKED ISSUE #2897: Test: SQL Server 2008 coverage
Our functional tests usually run on recent versions of LocalDB. We want to support older versions of SQL Server including SQL Server 2008 so we should figure out a way (through automation or not) to have good coverage for it.

### PR DESCRIPTION
**Pull request by natemcmaster**
Adds `[ConditionalFact]` and  `[ConditionalTheory]` to our testing framework (similar to aspnet/testing, but better :) )

Allows skipping tests based on test configuration.

Fix #2897 
~~Depends on #3011~~

### CODE REVIEW

**Comment by natemcmaster** [comment]
NB: Some SQL server tests cannot run as they test SEQUENCE functionality. These will always be ignored.
image

**Comment by anpete** [comment]
Nice!

**Comment by natemcmaster** [comment]
Update: #3011 has been merged.

**Comment by natemcmaster** [comment]
Ping! Review plz.

File: test/EntityFramework.Core.FunctionalTests/TestUtilities/Xunit/ConditionalAttributeDiscoverer.cs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+            IXunitTestCaseDiscoverer innerDiscoverer;
+            if (testMethod.Method.GetCustomAttributes(typeof(TheoryAttribute)).Any())
+            {
+                innerDiscoverer = new TheoryDiscoverer(_diagnosticMessageSink);
+            }
+            else
+            {
+                innerDiscoverer = new FactDiscoverer(_diagnosticMessageSink);
+            }
```
Comment (ajcvickers) [inline test/EntityFramework.Core.FunctionalTests/TestUtilities/Xunit/ConditionalAttributeDiscoverer.cs]: Consider ?:

File: test/EntityFramework.CrossStore.FunctionalTests/TestModels/CrossStoreContext.cs [hunk 2]
```diff
@@ -23,7 +24,12 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
                 .Entity<SimpleEntity>(eb =>
                     {
                         eb.Property(typeof(string), SimpleEntity.ShadowPartitionIdName);
-                        eb.Property(e => e.Id).UseSqlServerSequenceHiLo();
+
+                        if (TestEnvironment.GetFlag(nameof(SqlServerCondition.SupportsSequences)) ?? true)
+                        {
+                            eb.Property(e => e.Id).UseSqlServerSequenceHiLo();
+                        }
```
Comment (ajcvickers) [inline test/EntityFramework.CrossStore.FunctionalTests/TestModels/CrossStoreContext.cs]: Can we just always run this test without using sequences?

**Comment by ajcvickers** [comment]
:shipit:

File: test/EntityFramework.CrossStore.FunctionalTests/TestModels/CrossStoreContext.cs (hunk 2 above)
Comment (natemcmaster) [inline test/EntityFramework.CrossStore.FunctionalTests/TestModels/CrossStoreContext.cs]: Yes. Will update.

**Pull request merged**